checks/core/certificate_expiration.py

```diff
@@ -0,0 +1,68 @@
+"""
+SSL certificate should not expire for at least some minimum number of days.
+
+Returns expiration date.
+"""
+import asyncio
+import datetime
+import logging
+import ssl
+from urllib.parse import urlparse
+
+import cryptography
+import cryptography.x509
+from cryptography.hazmat.backends import default_backend as crypto_default_backend
+
+from poucave.typings import CheckResult
+from poucave.utils import utcnow
+
+
+logger = logging.getLogger(__name__)
+
+
+EXPOSED_PARAMETERS = ["url", "min_remaining_days"]
+
+
+# Bound the alert thresholds.
+LOWER_MIN_REMAINING_DAYS = 7
+UPPER_MIN_REMAINING_DAYS = 60  # No need to warn more than 2 months in advance.
+
+
+async def fetch_cert(url):
+    parsed = urlparse(url)
+    address = (parsed.netloc, parsed.port or 443)
+    loop = asyncio.get_event_loop()
+    cert_pem = await loop.run_in_executor(
+        None, lambda: ssl.get_server_certificate(address)
+    )
+    return cryptography.x509.load_pem_x509_certificate(
+        cert_pem.encode("utf8"), backend=crypto_default_backend()
+    )
+
+
+async def run(
+    url: str,
+    percentage_remaining_validity: int = 10,
+    min_remaining_days: int = LOWER_MIN_REMAINING_DAYS,
+) -> CheckResult:
+    cert = await fetch_cert(url)
+    start = cert.not_valid_before.replace(tzinfo=datetime.timezone.utc)
+    end = cert.not_valid_after.replace(tzinfo=datetime.timezone.utc)
+    lifespan = (end - start).days
+
+    # The minimum remaining days depends on the certificate lifespan.
+    relative_minimum = lifespan * percentage_remaining_validity / 100
+    bounded_minimum = int(
+        min(UPPER_MIN_REMAINING_DAYS, max(min_remaining_days, relative_minimum))
+    )
+    remaining_days = (end - utcnow()).days
```

Comment:
I wanted to make sure I understood this bit.

A Let's Encrypt cert has a 90 day lifespan. That means with the defaults:

* Relative minimum is 9 days
* Bounded minimum is 9 days

We often provision 1-year certs (365 days)

* Relative minimum is 37 days
* Bounded minimum is 37 days

If we pay for a 3-year cert (1095 days)

* Relative minimum is 110 days
* Bounded minimum is 60 days

That seems fine to me, but potentially confusing if the caller configures a smaller window (like 7 days) and the relative minimum is picked instead. But, if they get the alert, they should spend their time replacing the cert, not figuring out the config...

Reply:
Thanks for the prompt review!

> they should spend their time replacing the cert, not figuring out the config...

Hehe true! 

Maybe the default values are not adequate. With `min_remaining_days = 15` and `percentage_remaining_validity = 5`, `UPPER_MIN_REMAINING_DAYS = 30` we have:

* 90 days lifespan: relative is 5 days, bounded is 15 days
* 365 days lifespan: relative is 18 days, bounded is 18 days
* 1095 days lifespan: relative is 55 days, bounded is 30 days